### PR TITLE
feat: lock existing position and sfs

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #XXX

## What is the purpose of the change

We do not have a flow for locking existing positions and superfluid staking them. This PR adds this functionality.

## Testing and Verifying

Both msg server and keeper method have new tests added.

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/superfluid/keeper/stake.go [hunk 1]
```diff
@@ -662,3 +664,49 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
 		fn(index+int64(i), delegation)
 	}
 }
+
+func (k Keeper) LockExistingFullRangePositionAndSFStake(ctx sdk.Context, positionId uint64, sender sdk.AccAddress, valAddr string) (concentratedLockID uint64, err error) {
```
Comment (stackman27) [inline x/superfluid/keeper/stake.go:668]: just curious: what's the full context behind this message? Is it so that we can stake CL positions and get staking rewards?

Comment (stackman27) [reply x/superfluid/keeper/stake.go:668]: if so, then is there a reason why it is only for full range position? and should we remove the locks when user undelegates?

File: proto/osmosis/superfluid/tx.proto [hunk 2]
```diff
@@ -225,4 +229,16 @@ message MsgAddToConcentratedLiquiditySuperfluidPositionResponse {
     (gogoproto.nullable) = false
   ];
   uint64 lock_id = 4 [ (gogoproto.moretags) = "yaml:\"lock_id\"" ];
+}
+
+// ===================== MsgLockExistingFullRangePositionAndSFStake
+message MsgLockExistingFullRangePositionAndSFStake {
+  uint64 position_id = 1 [ (gogoproto.moretags) = "yaml:\"position_id\"" ];
+  string sender = 2 [ (gogoproto.moretags) = "yaml:\"sender\"" ];
```
Comment (stackman27) [inline proto/osmosis/superfluid/tx.proto:237]: isn't sender always position owner? wondering if we can just use `Position.Address`

File: x/superfluid/types/msgs.go [hunk 3]
```diff
@@ -392,3 +393,40 @@ func (msg MsgAddToConcentratedLiquiditySuperfluidPosition) GetSigners() []sdk.Ac
 	}
 	return []sdk.AccAddress{sender}
 }
+
```
Comment (stackman27) [inline x/superfluid/types/msgs.go:396]: is adding CLI going to be in a different PR?

File: x/superfluid/keeper/stake.go (hunk 1 above)
Comment (czarcas7ic) [reply x/superfluid/keeper/stake.go:668]: The only use case for locking in the context of CL is to SFS. The context behind this message is that we currently only allow direct create and SFS CL position messages, which means if someone had a full range position and wanted to SFS it, they would need to remove the position and create a new one, which is not a great UX. We also can only SFS a full range position as of now, so locking any other position (non full range) would not be a valid action.

File: proto/osmosis/superfluid/tx.proto (hunk 2 above)
Comment (mattverse) [reply proto/osmosis/superfluid/tx.proto:237]: No actually, this field would get automatically get generated thus can be deleted

File: x/superfluid/keeper/stake.go (hunk 1 above)
Comment (mattverse) [inline x/superfluid/keeper/stake.go:668]:
```suggestion
// LockExistingFullRangePositionAndSFStake locks existing position and superfluid stakes the position.
// This method would error if the existing position is not a full ranged position.
func (k Keeper) LockExistingFullRangePositionAndSFStake(ctx sdk.Context, positionId uint64, sender sdk.AccAddress, valAddr string) (concentratedLockID uint64, err error) {
```

File: x/superfluid/keeper/stake_test.go [hunk 4]
```diff
@@ -1188,6 +1192,173 @@ func (suite *KeeperTestSuite) TestPartialSuperfluidUndelegateToConcentratedPosit
...
+		},
+		{
+			name:          "position upper tick is not full range",
+			positionId:    2,
+			lowerTick:     cltypes.MinInitializedTick,
+			upperTick:     cltypes.MaxTick - 1,
+			expectedError: cltypes.PositionNotFullRangeError{PositionId: 2, LowerTick: cltypes.MinInitializedTick, UpperTick: cltypes.MaxTick - 1},
+		},
+	}
+
```
Comment (mattverse) [inline x/superfluid/keeper/stake_test.go:1240]: can we have test case for sender not the owner of position ?

File: x/superfluid/keeper/stake_test.go [hunk 5]
```diff
@@ -1188,6 +1192,173 @@ func (suite *KeeperTestSuite) TestPartialSuperfluidUndelegateToConcentratedPosit
...
+	// // Join the balancer pool.
+	// // Note the account balance before and after joining the pool.
+	// balanceBeforeJoin := bankKeeper.GetAllBalances(ctx, poolCreateAcc)
+	// _, _, err = gammKeeper.JoinPoolNoSwap(ctx, poolCreateAcc, balancerPooId, gammtypes.OneShare.MulRaw(50), sdk.Coins{})
+	// s.Require().NoError(err)
+	// balanceAfterJoin := bankKeeper.GetAllBalances(ctx, poolCreateAcc)
+
+	// // The balancer join pool amount is the difference between the account balance before and after joining the pool.
+	// joinPoolAmt, _ = balanceBeforeJoin.SafeSub(balanceAfterJoin)
```
Comment (mattverse) [inline x/superfluid/keeper/stake_test.go:1320]:
```suggestion
```

File: x/superfluid/keeper/stake_test.go [hunk 6]
```diff
@@ -1188,6 +1192,173 @@ func (suite *KeeperTestSuite) TestPartialSuperfluidUndelegateToConcentratedPosit
...
+				s.Require().Equal(synthLock.UnderlyingLockId, clLock.ID)
+				s.Require().Equal(synthLock.SynthDenom, keeper.StakingSyntheticDenom(clLock.Coins[0].Denom, valAddr.String()))
+				s.Require().Equal(synthLock.EndTime, time.Time{})
+
+				expAcc := types.NewSuperfluidIntermediaryAccount(clLock.Coins[0].Denom, valAddr.String(), 0)
+
+				// Check lockID connection with intermediary account
+				intAcc := s.App.SuperfluidKeeper.GetLockIdIntermediaryAccountConnection(s.Ctx, clLock.ID)
+				s.Require().Equal(intAcc.String(), expAcc.GetAccAddress().String())
+			}
```
Comment (mattverse) [inline x/superfluid/keeper/stake_test.go:1288]: what happens if you withdraw the cl position after?

File: x/superfluid/keeper/stake_test.go [hunk 7]
```diff
@@ -1188,6 +1192,173 @@ func (suite *KeeperTestSuite) TestPartialSuperfluidUndelegateToConcentratedPosit
...
+				// check synthetic lockup creation
+				synthLock, err := s.App.LockupKeeper.GetSyntheticLockup(s.Ctx, clLock.ID, keeper.StakingSyntheticDenom(clLock.Coins[0].Denom, valAddr.String()))
+				s.Require().NoError(err)
+				s.Require().Equal(synthLock.UnderlyingLockId, clLock.ID)
+				s.Require().Equal(synthLock.SynthDenom, keeper.StakingSyntheticDenom(clLock.Coins[0].Denom, valAddr.String()))
+				s.Require().Equal(synthLock.EndTime, time.Time{})
+
+				expAcc := types.NewSuperfluidIntermediaryAccount(clLock.Coins[0].Denom, valAddr.String(), 0)
+
+				// Check lockID connection with intermediary account
```
Comment (mattverse) [inline x/superfluid/keeper/stake_test.go:1285]: can we also check the actual delegated amount and see if it matches the liquidity in the position?

**Review by mattverse — COMMENTED** [review]
LGTM, left some minor comments that I would love to see before merge

**Review by mattverse — APPROVED** [review]

File: x/superfluid/keeper/stake_test.go (hunk 6 above)
Comment (czarcas7ic) [reply x/superfluid/keeper/stake_test.go:1292]: Added withdraw attempt after locking https://github.com/osmosis-labs/osmosis/pull/5745/commits/4d5545c9df8d3edefc652f85bb63e7cdd9578cb5

File: x/superfluid/keeper/stake_test.go (hunk 4 above)
Comment (czarcas7ic) [reply x/superfluid/keeper/stake_test.go:1252]: non owner test case added here https://github.com/osmosis-labs/osmosis/pull/5745/commits/d3c1b0fb39150a80643df73f7d4f0977ebe0628a

File: proto/osmosis/superfluid/tx.proto (hunk 2 above)
Comment (czarcas7ic) [reply proto/osmosis/superfluid/tx.proto:237]: @mattverse is "sender" in the other protos here also unnecessary?

File: x/superfluid/types/msgs.go (hunk 3 above)
Comment (czarcas7ic) [reply x/superfluid/types/msgs.go:396]: added CLI here https://github.com/osmosis-labs/osmosis/pull/5745/commits/ccd45be0f984850c7ddaa11d5b62dd0b93542c04

File: x/superfluid/keeper/stake.go (hunk 1 above)
Comment (stackman27) [reply x/superfluid/keeper/stake.go:670]: sounds good, can we please add this in the comment?

**Review by stackman27 — APPROVED** [review]
LGTM! 🇺🇸

**Review by p0mvn — APPROVED** [review]
Only did a cursory review since there are 2 approvals already. LGTM
